File: 24-swap-nodes-in-pairs/swap-nodes-in-pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# prev.next = curr.next
# curr.next = prev.next.next
# prev.next.next = curr
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        i = 1
        curr = head
        prev = head
        
        while(curr != None):
            if(curr == head):
                temp = curr.next
                if(temp != None):
                    curr.next = curr.next.next
                    temp.next = curr
                    head = temp
            else:
                if(curr.next != None):
                    prev.next = curr.next
                    curr.next = prev.next.next
                    prev.next.next = curr
            
            prev = curr
            curr = curr.next
            
        return head
        

File: 24-swap-nodes-in-pairs/swap-nodes-in-pairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        // If we have one or less nodes return head
        if(head == null || head.next == null) return head;

        ListNode prevNode = head;
        ListNode leadNode = head.next;
        ListNode tailNode = leadNode;
        // Save the head node to be returned later
        ListNode returnNode = leadNode;

        while(leadNode != null){
           if(tailNode != returnNode) tailNode.next = leadNode;
           prevNode.next = leadNode.next;
           leadNode.next = prevNode;

           tailNode = prevNode;
           prevNode = prevNode.next;
           if(prevNode == null || prevNode.next == null) break;
           leadNode = prevNode.next;
        }
        return returnNode;
    }
}